### PR TITLE
feature: support rollback info compress

### LINKED ISSUE #3159: add rollback_info field unzip for undolog
## Why you need it?
 Is your feature request related to a problem? Please describe in details

When the amount of data that needs to be rolled back is very large, it is very slow to read undolog from the database, we should optimize it as much as possible
当需要回滚的数据量非常大时，从数据库中读取undolog是非常慢的，我们要尽可能的优化他

## How it could be?
A clear and concise description of what you want to happen. You can explain more about input of the feature, and output of it.

### PR DESCRIPTION
**Pull request by caohdgege**
### Ⅰ. Describe what this PR did
support rollback info compress
支持对undo_log中的rollback_info进行压缩存储

顺便把 OracleUndoLogManager 和 PostgresqlUndoLogManager 中的hard code用常量替换了

### Ⅱ. Does this pull request fix one issue?

fixes #3159

### Ⅲ. Why don't you add test cases (unit test/integration test)? 
already add

### CODE REVIEW

**Comment by caohdgege** [comment]
zip(default)和gzip两种压缩方式的压缩比例都挺高的，然后总的压缩+插入时间也比较短
lz4不适合，虽然压缩的快，但是压缩比例不高，这样压缩时间+插入undo_log的时间会比其他压缩方式长，不值得
bzip2和7z也不太合适，压缩用时太长，比别人压缩+插入的时间都长了。。。

**Comment by caohdgege** [comment]
image
image

**Comment by a364176773** [comment]
good job

**Comment by caohdgege** [comment]
ps: 以上测试时间数据都是在MySQL下测试得来，pg和Oracle只是进行了功能测试，没有进行性能测试。

**Comment by caohdgege** [comment]
补一下测试的时候的日志和统计数据。
[测试日志.zip](https://github.com/seata/seata/files/5355586/default.zip)

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/oracle/OracleUndoLogManager.java [hunk 1]
```diff
@@ -67,34 +73,37 @@ public int deleteUndoLogByLogCreated(Date logCreated, int limitRows, Connection
...
+                                           CompressorType compressType, Connection conn) throws SQLException {
+        insertUndoLog(xid, branchId,rollbackCtx, undoLogContent, compressType, State.Normal, conn);
     }
 
     @Override
     protected byte[] getRollbackInfo(ResultSet rs) throws SQLException {
         Blob b = rs.getBlob(ClientTableColumnsName.UNDO_LOG_ROLLBACK_INFO);
         byte[] rollbackInfo = BlobUtils.blob2Bytes(b);
-        return rollbackInfo;
+        CompressorType compressType = CompressorType.getByCode(rs.getInt(ClientTableColumnsName.UNDO_LOG_COMPRESS_TYPE));
```
Comment (a364176773) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/oracle/OracleUndoLogManager.java:85]: Why not reduce the number of duplicate lines of code by putting decompression into AbstractUndoLogManager?

**Comment by a364176773** [comment]
config.txt application.properties application.properties supplement the related configuration

File: server/src/main/resources/file.conf.example [hunk 2]
```diff
@@ -83,6 +83,14 @@ server {
     logSaveDays = 7
     #schedule delete expired undo_log in milliseconds
     logDeletePeriod = 86400000
+    compress {
```
Comment (a364176773) [inline server/src/main/resources/file.conf.example:86]: 应该补充到script文件夹下的client中的conf里的file.conf文件

File: server/src/main/resources/file.conf.example [hunk 3]
```diff
@@ -83,6 +83,14 @@ server {
...
+      enable = true
+      // allow zip, gzip, 7z, lz4, bzip2, default is zip
```
Comment (wangliang181230) [inline server/src/main/resources/file.conf.example:88]: `#`, not `//`

File: seata-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 4]
```diff
@@ -150,6 +150,24 @@
...
       "defaultValue": true
     },
+    {
+      "name": "seata.client.undo.compress.enable",
+      "type": "java.lang.Boolean",
+      "sourceType": "io.seata.spring.boot.autoconfigure.properties.client.UndoCompressProperties",
+      "defaultValue": true
+    },
+    {
+      "name": "seata.client.undo.compress.type",
```
Comment (wangliang181230) [inline seata-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json:160]: Please add optional values for this configuration in `hints` below. like `seata.transport.compressor`

Comment (caohdgege) [reply seata-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json:160]: done

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/oracle/OracleUndoLogManager.java (hunk 1 above)
Comment (caohdgege) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/oracle/OracleUndoLogManager.java:85]: done

File: common/src/main/java/io/seata/common/util/SizeUtil.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        long number;
+        try {
+            number = NumberUtils.toLong(size2Lower.substring(0, size.length() - 1));
+        } catch (NumberFormatException | NullPointerException ex) {
+            throw new IllegalArgumentException("could not convert '" + size + "' to byte length");
+        }
+
+        switch (unit) {
+            case 'k':
+                return number * 1024;
```
[the remark this answers is not in the document]
Comment (caohdgege) [reply common/src/main/java/io/seata/common/util/SizeUtil.java]: done

File: script/client/at/db/mysql.sql [hunk 6]
```diff
@@ -5,6 +5,7 @@ CREATE TABLE IF NOT EXISTS `undo_log`
     `xid`           VARCHAR(100) NOT NULL COMMENT 'global transaction id',
     `context`       VARCHAR(128) NOT NULL COMMENT 'undo_log context,such as serialization',
     `rollback_info` LONGBLOB     NOT NULL COMMENT 'rollback info',
+    `compress_type` TINYINT      NOT NULL COMMENT 'compress type',
```
Comment (jsbxyyx) [inline script/client/at/db/mysql.sql:8]: compress_type should be placed in context field.

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java [hunk 7]
```diff
@@ -392,5 +414,18 @@ protected abstract void insertUndoLogWithNormal(String xid, long branchId, Strin
      * @return
      * @throws SQLException
      */
-    protected abstract byte[] getRollbackInfo(ResultSet rs) throws SQLException;
+    protected byte[] getRollbackInfo(ResultSet rs) throws SQLException  {
+        byte[] rollbackInfo = rs.getBytes(ClientTableColumnsName.UNDO_LOG_ROLLBACK_INFO);
+        CompressorType compressType = CompressorType.getByCode(rs.getInt(ClientTableColumnsName.UNDO_LOG_COMPRESS_TYPE));
+        return CompressorFactory.getCompressor(compressType.getCode()).decompress(rollbackInfo);
```
Comment (jsbxyyx) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java:420]: rollbackInfo needs to determine whether to decompress according to the compression type value in the context.

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java [hunk 8]
```diff
@@ -213,11 +228,17 @@ public void flushUndoLogs(ConnectionProxy cp) throws SQLException {
...
+            compressorType = ROLLBACK_INFO_COMPRESS_TYPE;
+            undoLogContent = CompressorFactory.getCompressor(compressorType.getCode()).compress(undoLogContent);
+        }
+
         if (LOGGER.isDebugEnabled()) {
             LOGGER.debug("Flushing UNDO LOG: {}", new String(undoLogContent, Constants.DEFAULT_CHARSET));
         }
 
-        insertUndoLogWithNormal(xid, branchId, buildContext(parser.getName()), undoLogContent,
+        insertUndoLogWithNormal(xid, branchId, buildContext(parser.getName()), undoLogContent, compressorType,
```
Comment (jsbxyyx) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java:241]: compressorType placed in context.

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java (hunk 7 above)
Comment (caohdgege) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java]: ok, move to context

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java [hunk 9]
```diff
@@ -392,5 +413,22 @@ protected abstract void insertUndoLogWithNormal(String xid, long branchId, Strin
      * @return
      * @throws SQLException
      */
-    protected abstract byte[] getRollbackInfo(ResultSet rs) throws SQLException;
+    protected byte[] getRollbackInfo(ResultSet rs) throws SQLException  {
+        byte[] rollbackInfo = rs.getBytes(ClientTableColumnsName.UNDO_LOG_ROLLBACK_INFO);
+
+        String rollbackInfoContext = rs.getString(ClientTableColumnsName.UNDO_LOG_CONTEXT);
+        Map<String, String> context = CollectionUtils.decodeMap(rollbackInfoContext);
+        CompressorType compressorType = CompressorType.getByName(context.get(UndoLogConstants.COMPRESSOR_TYPE_KEY));
```
Comment (jsbxyyx) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java:421]: old version upgrade new version. compressorType may be not exists in database. need to consider?

Comment (caohdgege) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java]: use Map#getOrDefault to compatible history data

File: common/src/main/java/io/seata/common/util/SizeUtil.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+            case 'k':
+                return number * RADIX;
+            case 'm':
+                return number * RADIX * RADIX;
+            case 'g':
+                return number * RADIX * RADIX * RADIX;
+            case 't':
+                return number * RADIX * RADIX * RADIX * RADIX;
+            default:
+                throw new IllegalArgumentException("could not convert " + size + " to byte length");
```
Comment (mzzw) [inline common/src/main/java/io/seata/common/util/SizeUtil.java:57]: is the `size` need wrapped with `'`?

Comment (caohdgege) [reply common/src/main/java/io/seata/common/util/SizeUtil.java]: ok

**Review by a364176773 — APPROVED** [review]
LGTM

File: core/src/main/java/io/seata/core/constants/ClientTableColumnsName.java [hunk 11]
```diff
@@ -22,6 +22,12 @@
  */
 public interface ClientTableColumnsName {
 
+    /**
+     * The constant undo_log column name xid
+     * this field is not use in mysql
+     */
+    String UNDO_LOG_ID = "id";
```
Comment (jsbxyyx) [inline core/src/main/java/io/seata/core/constants/ClientTableColumnsName.java:29]: why add id?

Comment (caohdgege) [reply core/src/main/java/io/seata/core/constants/ClientTableColumnsName.java:29]: 为了让 OracleUndoLogManager和PostgresUndoLogManager 里面的SQL不用hard code

**Review by jsbxyyx — APPROVED** [review]
LGTM

**Pull request merged**